feat(acquisition): add manual acquisition

krohrsb opened the pull request:
This adds the notion of a `manualAcquisition` in order to facilitate manual specification of acquisitions outside of containers specifically.

`/var/log` is already a mapped volume, so theoretically you can specify a manual acquis for any other log file of your desire, such as `auth.log` rather than only being able to parse container logs.

This is an attempt to solve for #88 -- though I made assumptions on naming conventions and such and am completely open to altering it, if the idea in general is accepted.

Here is an example:

Using manual acquis + some dummy normal acquis:

```yaml
  additionalAcquisition:
    - filenames:
        - /var/log/auth.log
      labels:
        type: syslog
        program: sshd
  acquisition:
    # -- Specify each pod you want to process it logs (namespace, podName and program)
    - namespace: "foo" #ingress-nginx
      # -- to select pod logs to process
      podName: "foo-*" #ingress-nginx-controller-*
      # -- program name related to specific parser you will use (see https://hub.crowdsec.net/author/crowdsecurity/configurations/docker-logs)
      program: "foo" #nginx
```

Output:

```yaml
# Source: crowdsec/templates/acquis-configmap.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: acquis-configmap
data:
  acquis.yaml: |-
    ---
    filenames:
      - /var/log/containers/foo-*_foo_*.log
    force_inotify: true
    labels:
      type: docker
      program: foo
    ---
    filenames:
    - /var/log/auth.log
    labels:
      program: sshd
      type: syslog
```

he2ss commented on charts/crowdsec/values.schema.json:
```diff
@@ -26,13 +26,50 @@
             ],
             "title": "Agent"
         },
+        "ManualAcquisition": {
```
For the validation, I think it's more complex than that, if you take a look at the [available datasources](https://docs.crowdsec.net/docs/next/data_sources/intro). Each one has it own parameters.

Maybe ignoring this part for now, or you can build the complex schema following condition on the `source`

he2ss commented on charts/crowdsec/values.yaml:
```diff
@@ -198,6 +198,12 @@ lapi:
 
 # agent will deploy pod on every node as daemonSet to read wanted pods logs
 agent:
+  # manualAcquisition:
```
Better if the name is `additionalAcquisition` no ?

he2ss reviewed (commented):
Hi @krohrsb,

Thanks for your contribution. Indeed it's a good Idea to have additional custom configuration on the acquisition.

I let you few comments.

krohrsb commented:
Thanks for the review I'll see about getting the suggestions implemented.

krohrsb commented:
I've gone ahead and renamed it to `additionalAcquisition` and added JSON schema entries for all data sources. Though if you rather we left that out for now and did a generic catchall, we can do that as well, just let me know.

he2ss commented on charts/crowdsec/values.schema.json:
```diff
@@ -99,6 +134,519 @@
+                "brokers"
+            ]
+        },
+        "WinEventLogSource": {
```
There is an issue with this validation, if you take a look at the [xpath doc](https://docs.crowdsec.net/docs/next/data_sources/windows_evt_log#xpath_query).

he2ss replied:
Sorry, the schema is good. I had an issue because I had a typo in my values.

he2ss commented on charts/crowdsec/values.yaml:
```diff
@@ -198,6 +198,16 @@ lapi:
 
 # agent will deploy pod on every node as daemonSet to read wanted pods logs
 agent:
```
you can add a documentation comment above `additionalAcquisition` with a link to the datasources. Somethink like :

```
# -- To add custom acquisitions using available datasources (https://docs.crowdsec.net/docs/next/data_sources/intro)
```

he2ss reviewed (commented):
Hi @krohrsb,

Still some little remarks and also an advice. You can use lint to test this validation schema, you can additional config to `charts/crowdsec/ci/crowdsec-values.yaml`.

here is the example below I used to spot the bug in the schema definition `WinEventLogSource`:

```
  additionalAcquisition:
    - source: file
      filenames:
        - /var/log/auth.log
      force_inotify: true
      labels:
        type: syslog
    - source: file
      filename:
        - /var/log/auth.log
      labels:
        type: syslog
    - source: journalctl
      journalctl_filter:
        - "_SYSTEMD_UNIT=sshd.service"
      labels:
        type: syslog
    - source: cloudwatch
      group_name: /aws/my/group
      aws_profile: monitoring
      stream_regexp: "^stream[0-9]+$"
      labels:
        type: apigateway
    - source: kinesis
      stream_arn: arn:aws:kinesis:eu-west-1:123456789012:stream/my-stream
      use_enhanced_fanout: true
      consumer_name: my-consumer
      labels:
        type: mytype
    - source: syslog
      listen_addr: 127.0.0.1
      listen_port: 514
      labels:
        type: syslog
        program: syslog
    - source: docker
      container_name:
        - "my-container"
      container_id:
        - "1234567890abcdef"
      labels:
        type: docker
    - source: wineventlog
      xpath_query: |
          <QueryList><Query><Select Path=\"Security\">*[System[(EventID=42) and (Level=2)]]</Select></Query></QueryList>
      labels:
      type: eventlog
    - source: kafka
      brokers:
        - "localhost:9093"
      topic: "my-topic"
      timeout: 5
      tls:
        insecure_skip_verify: true
        client_cert: /path/kafkaClient.certificate.pem
        client_key: /path/kafkaClient.key
        ca_cert: /path/ca.crt
      labels:
        type: nginx
```

he2ss commented:
@krohrsb Thanks !

teceP commented:
Hi @krohrsb, thank you for creating this charts!

Can we add "appsec" as source? 
Id like to add this config to the acquis.yaml:
```yaml
listen_addr: 127.0.0.1:7422
appsec_config: crowdsecurity/appsec-default
name: myAppSecComponent
source: appsec
labels:
  type: appsec
```
as quoted in https://docs.crowdsec.net/docs/next/appsec/installation/

When i try 
```yaml
agent:
    additionalAcquisition:
      - source: appsec
        listen_addr: 127.0.0.1:7422
        path: /
        appsec_config: crowdsecurity/appsec-default
        labels:
          type: appsec
```

I get following schema validation errors:
```yaml
Error: values don't meet the specifications of the schema(s) in the following chart(s):
crowdsec:
- agent.additionalAcquisition.1: Must validate at least one schema (anyOf)
- agent.additionalAcquisition.1: Additional property listen_addr is not allowed
- agent.additionalAcquisition.1: Additional property path is not allowed
- agent.additionalAcquisition.1: Additional property appsec_config is not allowed
- agent.additionalAcquisition.1.source: agent.additionalAcquisition.1.source does not match: "kinesis"
```

comminutus commented:
@teceP were you ever able to get AppSec working using the helm chart?

teceP commented:
@comminutus unfortunately not, but I didnt tried for a long time.
